File: DSA/sorting_algorithms/insertion_sort/python/insertion_sort.py
#!/bin/python3
"""
Insertion Sort Algorithm Implementation in Python

This algorithm builds the sorted array one item at a time.
It is efficient for small datasets and nearly sorted arrays.

Time Complexity:
- Best: O(n)
- Average: O(n^2)
- Worst: O(n^2)

Space Complexity: O(1)
Stable: Yes
"""

def insertion_sort(arr):
    """
    Sorts a list in ascending order using the insertion sort algorithm.

    Args:
        arr (list): The list of elements to sort.

    Returns:
        list: Sorted list in ascending order.
    """
    for i in range(1, len(arr)):
        current_value = arr[i]
        j = i - 1

        while j >= 0 and current_value < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1

        arr[j + 1] = current_value

    return arr

# Example usage
if __name__ == "__main__":
    data = [10, 3, 0, 2, 6, 1, 5, 4, 8, 7, 9]
    print("Original array:", data)
    sorted_data = insertion_sort(data)
    print("Sorted array:", sorted_data)

File: DSA/sorting_algorithms/insertion_sort/javascript/insertionSort.js
#!/usr/bin/env node
/**
 * Insertion Sort Algorithm Implementation in JavaScript
 *
 * Builds the final sorted array one element at a time.
 * Efficient for small or nearly sorted datasets.
 *
 * Time Complexity:
 * - Best: O(n)
 * - Average: O(n^2)
 * - Worst: O(n^2)
 *
 * Space Complexity: O(1)
 * Stable: Yes
 */

/**
 * Sorts an array in ascending order using insertion sort.
 * @param {number[]} arr - The array to sort.
 * @returns {number[]} Sorted array.
 */
function insertionSort(arr) {
  for (let i = 1; i < arr.length; i++) {
    const currentValue = arr[i];
    let j = i - 1;

    while (j >= 0 && currentValue < arr[j]) {
      arr[j + 1] = arr[j];
      j--;
    }

    arr[j + 1] = currentValue;
  }
  return arr;
}

// Example usage
const data = [10, 3, 0, 2, 6, 1, 5, 4, 8, 7, 9];
console.log("Original array:", data);
console.log("Sorted array:", insertionSort([...data]));
